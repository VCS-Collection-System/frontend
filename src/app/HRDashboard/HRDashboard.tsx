import * as React from 'react';
import { Button, Card, CardBody, CardFooter, Flex, FlexItem, Grid, GridItem } from '@patternfly/react-core';
import { HRInbox } from './AgencyInbox/HRInbox';
import { HRPriority } from './PriorityInbox/HRPriority';
import { HRSearch } from './Search/HRSearch';
import { useHistory } from 'react-router-dom';
import Feature from '@app/Components/Feature';

const HRDashboard: React.FunctionComponent = () => {
  const history = useHistory();

  return (
    <Card isRounded>
      <CardBody>
        <Grid hasGutter>
          <Feature name='hrSearch'
            renderOff={() =>
              <>
                <GridItem span={6}>
                  <HRInbox />
                </GridItem>
                <GridItem span={6}>
                  <HRPriority />
                </GridItem>
              </>
            }>
            <>
              <GridItem span={4}>
                <HRInbox />
              </GridItem>
              <GridItem span={4}>
                <HRSearch />
              </GridItem>
              <GridItem span={4}>
                <HRPriority />
              </GridItem>
            </>
          </Feature>
        </Grid>
      </CardBody>
      <CardFooter>
        <Flex>
          <FlexItem align={{ default: "alignLeft" }}>
            <Button 
              variant="plain" 
              onClick={() => history.push("/attestmenu")
            }>
              Back
            </Button>
          </FlexItem>
        </Flex>
      </CardFooter>
    </Card>
  )
};

export { HRDashboard };

