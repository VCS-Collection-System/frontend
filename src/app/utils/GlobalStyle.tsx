import * as React from 'react';
import { createGlobalStyle } from 'styled-components';

const GlobalStyle = () => {

  const fontColor = "rgb(180, 180, 178)"; 
  const backgroundColor200 = "var(--pf-global--BackgroundColor--dark-200)";
  const backgroundColor300 = "var(--pf-global--BackgroundColor--dark-300)";
  const backgroundColor400 = "var(--pf-global--BackgroundColor--dark-400)"; 
  const borderColor200 = "var(--pf-global--BorderColor--200)"; 

  const GlobalStyles = createGlobalStyle`
    .pf-c-page__main {
      background-color: ${backgroundColor400}
    }
    .pf-c-button.pf-m-plain#span {
      color: ${fontColor};
    }
    .pf-c-title.pf-m-3xl { 
      color: ${fontColor};
    }
    .pf-c-card {
      --pf-c-card--BackgroundColor: ${backgroundColor200};
      color: ${fontColor};
    }
    .pf-c-card .pf-c-card{
      --pf-c-card--BackgroundColor: ${backgroundColor400};
      color: ${fontColor};
    }
    .pf-c-card .pf-c-card .pf-c-card{
      --pf-c-card--BackgroundColor: ${backgroundColor200};
      color: ${fontColor};
    }
    .pf-c-form-control:disabled {
      --pf-c-form-control--BackgroundColor: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-form-control {
      --pf-c-form-control--BackgroundColor: ${backgroundColor300};
      --pf-c-form-control--BorderTopColor: ${backgroundColor300};
      --pf-c-form-control--BorderRightColor: ${backgroundColor300};
      --pf-c-form-control--BorderBottomColor: ${backgroundColor300};
      --pf-c-form-control--BorderLeftColor: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-button.pf-m-tertiary {
      color: ${fontColor};
    };
    .pf-c-button.pf-m-control {
      background-color: ${backgroundColor300};
      --pf-c-button--after--BorderColor: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-calendar-month {
      --pf-c-calendar-month--BackgroundColor: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-calendar-month__dates-cell::before {
      background-color: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-calendar-month__date {
      color: ${fontColor};
    };
    .pf-c-select__toggle {
      background-color: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-form-control[readonly] {
      background-color: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-form__helper-text {
      color: ${fontColor};
    };
    .pf-c-menu {
      --pf-c-menu--BackgroundColor: ${backgroundColor300};
      color: color: ${fontColor};
    };
    .pf-c-menu__item {
      background-color: ${backgroundColor300};
      color: ${fontColor};
    }
    .pf-c-input-group {
      --pf-c-input-group--BackgroundColor: ${backgroundColor300};
    }
    .pf-c-search-input {
      --pf-c-search-input__text--before--BorderColor: ${borderColor200};
      --pf-c-search-input__text--after--BorderBottomColor: ${borderColor200};
      color: ${fontColor};
    }
    .pf-c-search-input__text-input {
      color: ${fontColor};
    }
    .pf-c-modal-box {
      --pf-c-modal-box--BackgroundColor: ${backgroundColor300};
      color: ${fontColor};
    }
    .pf-c-radio {
      --pf-c-radio__label--Color: ${fontColor};
    }
    .pf-c-button.pf-m-plain {
      color: ${fontColor};
    }
    `;

    return <GlobalStyles />
};

export default GlobalStyle;