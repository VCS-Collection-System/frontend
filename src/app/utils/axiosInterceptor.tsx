import axios from "axios";

const axiosInterceptor = axios.create({})

axiosInterceptor.interceptors.request.use(
    config => {
        config.headers['Authorization'] = 'Bearer ' + sessionStorage.getItem("token");
        return config;
    },
    error => {
        Promise.reject(error)
    });
export default axiosInterceptor;