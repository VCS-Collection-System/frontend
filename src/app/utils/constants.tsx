declare global {  
  interface Window {
    REACT_APP_PAM_URL_BASE: string;
    REACT_APP_PAM_CONTAINER_ID: string;
    REACT_APP_KEYCLOAK_COUNTRY: string;
    REACT_APP_KEYCLOAK_EMAIL: string;
    REACT_APP_KEYCLOAK_FIRST_NAME: string;
    REACT_APP_KEYCLOAK_ID: string;
    REACT_APP_KEYCLOAK_LAST_NAME: string;
    REACT_APP_MAINTENANCE_BANNER: boolean;
    REACT_APP_MAINTENANCE_BANNER_TEXT: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    REACT_APP_FEATURE_FLAGS: Array<Record<string, any>>;
    REACT_APP_MAX_DOSES: Record<string, number>;
    REACT_APP_CONSENT_LABEL: string;
    REACT_APP_ALTERNATE_CONSENT_LABEL: string;
    REACT_APP_BLOCKED_COUNTRIES: Array<string>;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    REACT_APP_GLOBAL_CONFIG: Record<string, any>;
  }
}

// PAM urls
export const PAM_BASE_URL = window.REACT_APP_PAM_URL_BASE; 
export const SUBMIT_VAX_URL = PAM_BASE_URL + '/attestation/vax';
export const SUBMIT_TEST_URL = PAM_BASE_URL + '/attestation/covid-test-result';
export const PAM_TASK_URL = PAM_BASE_URL + '/tasks/';
export const RESERVED_TASK_URL = PAM_TASK_URL + 'owners?status=Reserved&user=';
export const INPROGRESS_TASK_URL = PAM_TASK_URL + 'owners?status=InProgress&user=';
export const ALL_TASK_URL = PAM_TASK_URL + 'owners?status=Ready&pageSize=200';
export const PRIORITY_URL = PAM_BASE_URL + '/query/priority-results';
export const PRIORITY_CONFIRM_URL = PAM_BASE_URL + '/query/priority-result/';
export const ATTACHMENT_URL = PAM_BASE_URL + '/attestation/attachment';
export const EMPLOYEE_URL = PAM_BASE_URL + '/query/employee/';
export const EMPLOYEE_NAME_URL = PAM_BASE_URL + '/query/employee-name/';
export const EMPLOYEE_VAX_HISTORY_URL = PAM_BASE_URL + '/query/vax-document/';
export const EMPLOYEE_TEST_HISTORY_URL = PAM_BASE_URL + '/query/test-result-document/';
export const VAX_TASK_DOCUMENT_URL = PAM_BASE_URL + '/query/task/document/';
export const DELETE_VAX_URL = PAM_BASE_URL + '/attestation/delete-vax-record/';
export const CONFIG_URL = PAM_BASE_URL + '/query/country-configs/';
export const SAVE_CONFIG_URL = PAM_BASE_URL + '/attestation/country-configs';

// List of countries allow to use this app
export const BLOCKED_COUNTRIES = window.REACT_APP_BLOCKED_COUNTRIES || []

export const KONAMI_CODE = [ "ArrowUp", "ArrowUp", "ArrowDown", "ArrowDown", "ArrowLeft", "ArrowRight", "ArrowLeft", "ArrowRight", "KeyB", "KeyA"];

// Maintenance Banner
export const MAINTENANCE_BANNER = window.REACT_APP_MAINTENANCE_BANNER || false
export const MAINTENANCE_BANNER_TEXT = window.REACT_APP_MAINTENANCE_BANNER_TEXT || ""

// Keycloak claims
export const KEYCLOAK_COUNTRY = window.REACT_APP_KEYCLOAK_COUNTRY
export const KEYCLOAK_EMAIL = window.REACT_APP_KEYCLOAK_EMAIL
export const KEYCLOAK_FIRST_NAME = window.REACT_APP_KEYCLOAK_FIRST_NAME
export const KEYCLOAK_ID = window.REACT_APP_KEYCLOAK_ID
export const KEYCLOAK_LAST_NAME = window.REACT_APP_KEYCLOAK_LAST_NAME
export const FEATURE_FLAGS = window.REACT_APP_FEATURE_FLAGS || []

// Maximum doses allowed by vaccine manufacturer
export const MAX_DOSES = window.REACT_APP_MAX_DOSES || {}

export const CONSENT_LABEL = window.REACT_APP_CONSENT_LABEL
export const ALTERNATE_CONSENT_LABEL = window.REACT_APP_ALTERNATE_CONSENT_LABEL

// Global country config
export const GLOBAL_CONFIG = window.REACT_APP_GLOBAL_CONFIG || {};

