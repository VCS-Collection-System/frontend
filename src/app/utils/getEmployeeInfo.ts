import {
  KEYCLOAK_FIRST_NAME, 
  KEYCLOAK_LAST_NAME, 
  KEYCLOAK_EMAIL,
  KEYCLOAK_COUNTRY,
  KEYCLOAK_ID
} from "@app/utils/constants"; 

export function getEmployeeInfo(keycloak) {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const jp = require('jsonpath');
  const profile = keycloak.profile;
  const employeeFirstName = jp.query(profile, KEYCLOAK_FIRST_NAME)[0];
  const employeeLastName = jp.query(profile, KEYCLOAK_LAST_NAME)[0];
  const employeeEmail = jp.query(profile, KEYCLOAK_EMAIL)[0];
  const employeeCountry = jp.query(profile, KEYCLOAK_COUNTRY)[0];
  const employeeID = jp.query(profile, KEYCLOAK_ID)[0];

  const isApprover = () => {
    if(employeeCountry && 
      (keycloak.hasRealmRole("vcs-superapprover") || keycloak.hasRealmRole("vcs-approver-" + employeeCountry.toLowerCase()))){
        return true;
    }

    return false;
  }

  const isSuperApprover = () => {
    return keycloak.hasRealmRole("vcs-superapprover");
  }

  return {
    firstName: employeeFirstName, 
    lastName: employeeLastName,
    email: employeeEmail,
    country: employeeCountry,
    id: employeeID,
    isApprover: isApprover(),
    isSuperApprover: isSuperApprover()
  }
}
