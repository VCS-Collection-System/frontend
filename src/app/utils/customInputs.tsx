import { useEffect, useState } from "react";
import { KONAMI_CODE } from './constants'

const useInputEvent = () => {
  const [key, setKey] = useState(null);
  useEffect(() => {
    const keyDownHandler = ({ code }) => setKey(code);
    const keyUpHandler = () => setKey(null);
    window.addEventListener('keydown', keyDownHandler);
    window.addEventListener('keyup', keyUpHandler);
    return () => {
      window.removeEventListener("keydown", keyDownHandler);
      window.removeEventListener("keyup", keyUpHandler)
    }
  }, []);
  return key;
}

const useSecretCode = (secretCode) => {
  const [count, setCount] = useState(0);
  const [success, setSuccess] = useState(false);
  const key = useInputEvent();
  
  useEffect(() => {
    if (key == null) return;
    if (key !== secretCode[count]) {
      setCount(0);
      return;
    }
    
    setCount((state) => state + 1);
    if (count + 1 === secretCode.length) {
      setSuccess(!success);
      setCount(0);
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [key]); 
  
  return success;
};

export const useDarkMode = () => {
  return useSecretCode(KONAMI_CODE);
};
