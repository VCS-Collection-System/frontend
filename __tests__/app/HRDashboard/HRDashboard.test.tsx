import React from 'react';
import { HRDashboard } from '../../../src/app/HRDashboard/HRDashboard';
import { act, fireEvent, render, screen } from '@testing-library/react'
import reactRouterDom from "react-router-dom"
import mockKeycloak from '../../../__mocks__/keycloakMock'
import { ALL_TASK_URL, INPROGRESS_TASK_URL, RESERVED_TASK_URL } from '@app/utils/constants';

jest.mock('@app/utils/axiosInterceptor', () => ({
  get: (URL) => {
    let returnData; 
    if(URL === RESERVED_TASK_URL + "mockId"){
      returnData = {data: {"task-summary":[]}}
    }
    else if(URL === INPROGRESS_TASK_URL + "mockId"){
      returnData = {data: {"task-summary":[]}}
    }
    else if(URL === ALL_TASK_URL){
      returnData = {data: {"task-summary":[]}}
    }

    else{
      returnData = {data: []};
    }

    return Promise.resolve(returnData)
  }
}))

jest.mock('@react-keycloak/web', () => ({
  useKeycloak: jest.fn( ()=>({keycloak: mockKeycloak}))
})) 

let findValue = {
  active: true
};

const mockFind = jest.fn(() => {
  return findValue; 
})

jest.mock("@app/utils/constants", () => ({
  KEYCLOAK_FIRST_NAME: "firstName",
  KEYCLOAK_LAST_NAME: "lastName",
  KEYCLOAK_EMAIL: "email",
  KEYCLOAK_COUNTRY: "attributes.agency[0]",
  KEYCLOAK_ID: "attributes.workforceid[0]",
  APPROVER_ROLES: ["mock-approver"],
  CONSENT_COUNTRIES: ["AUS"],
  FEATURE_FLAGS: {
    find: () => mockFind()
  },
  GLOBAL_CONFIG: {"default": {
    "allowPdf": false,
    "dateOfBirth": true,
    "consentCheckbox": false,
    "covidTest": false,
    "customInput": true,
    "hrdashboard": true,
    "hrsearch": true,
    "proofType": {"cdc": false, "divoc": false, "greenPass": false},
    "recovery": false,
    "fakeFeature": true
  }}
}))

jest.mock("react-router-dom")
const mockPush = jest.fn()
reactRouterDom.useHistory = jest.fn().mockReturnValue({ push: mockPush })

describe('<HRDashboard />', ()=> {
  it('should be defined', () => {
    expect(HRDashboard).toBeDefined()
  })

  it("should match existing snapshot", async () => {
    await act(async () => {
      const { asFragment } = render(<HRDashboard />)
      expect(asFragment()).toMatchSnapshot()
    })
  })

  it("should render on with flag", async () => {
    await act(async () => {
      render(<HRDashboard />)
    })

    expect(screen.getByText('[No In-Progress Tasks]')).toBeTruthy();
    expect(screen.findByDisplayValue('Red Hat VCS Search')).toBeTruthy();
  })

  it("should render on with flag", async () => {
    findValue = {
      active: false
    }
    await act(async () => {
      render(<HRDashboard />)
    })

    expect(screen.getByText('[No In-Progress Tasks]')).toBeTruthy();
  })

  it("should render on with flag", async () => {
    await act(async () => {
      render(<HRDashboard />)
    })

    const backButton = screen.getByRole('button', {name: /Back/i})
    fireEvent.click(backButton)
    expect(mockPush).toHaveBeenCalledWith('/attestmenu')
  })
})