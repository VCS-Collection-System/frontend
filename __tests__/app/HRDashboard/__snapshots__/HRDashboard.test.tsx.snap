// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HRDashboard /> should match existing snapshot 1`] = `
<DocumentFragment>
  <article
    class="pf-c-card pf-m-rounded"
    data-ouia-component-id="OUIA-Generated-Card-1"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__body"
    >
      <div
        class="pf-l-grid pf-m-gutter"
      >
        <div
          class="pf-l-grid__item pf-m-6-col"
        >
          <article
            class="pf-c-card pf-m-rounded"
            data-ouia-component-id="OUIA-Generated-Card-2"
            data-ouia-component-type="PF4/Card"
            data-ouia-safe="true"
            id=""
          >
            <div
              class="pf-c-card__title"
            >
              <p
                class=""
                data-pf-content="true"
              >
                Inbox
              </p>
              <div
                class="pf-c-content"
              >
                <small
                  class=""
                  data-pf-content="true"
                >
                  Select an item below to review
                </small>
              </div>
            </div>
            <div
              aria-label="Global"
              class="pf-c-menu"
              data-ouia-component-id="OUIA-Generated-Menu-1"
              data-ouia-component-type="PF4/Menu"
              data-ouia-safe="true"
            >
              <div
                class="pf-c-menu__content"
                style="--pf-c-menu__content--Height: 350px;"
              >
                <section
                  class="pf-c-menu__group"
                >
                  <h1
                    class="pf-c-menu__group-title"
                    id=""
                  >
                    Your In-Progress Tasks
                  </h1>
                  <ul
                    class="pf-c-menu__list indentList"
                    id="in-progress-list"
                  >
                    <li
                      class="pf-c-menu__list-item pf-m-disabled"
                      tabindex="-1"
                    >
                      <button
                        aria-current="false"
                        class="pf-c-menu__item"
                        disabled=""
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="pf-c-menu__item-main"
                        >
                          <span
                            class="pf-c-menu__item-text"
                          >
                            [No In-Progress Tasks]
                          </span>
                        </span>
                      </button>
                    </li>
                  </ul>
                </section>
                <section
                  class="pf-c-menu__group"
                >
                  <h1
                    class="pf-c-menu__group-title"
                    id=""
                  >
                    All Available Tasks
                  </h1>
                  <ul
                    class="pf-c-menu__list indentList"
                    id="all-available-list"
                  >
                    <li
                      class="pf-c-menu__list-item pf-m-disabled"
                      tabindex="-1"
                    >
                      <button
                        aria-current="false"
                        class="pf-c-menu__item"
                        disabled=""
                        tabindex="-1"
                        type="button"
                      >
                        <span
                          class="pf-c-menu__item-main"
                        >
                          <span
                            class="pf-c-menu__item-text"
                          >
                            [No Available Tasks]
                          </span>
                        </span>
                      </button>
                    </li>
                  </ul>
                </section>
              </div>
            </div>
          </article>
        </div>
        <div
          class="pf-l-grid__item pf-m-6-col"
        >
          <article
            class="pf-c-card pf-m-rounded warningText"
            data-ouia-component-id="OUIA-Generated-Card-3"
            data-ouia-component-type="PF4/Card"
            data-ouia-safe="true"
            id=""
          >
            <div
              class="pf-c-card__title"
            >
              <p
                class=""
                data-pf-content="true"
              >
                Priority Inbox
              </p>
              <div
                class="pf-c-content"
              >
                <small
                  class=""
                  data-pf-content="true"
                >
                  Select an item below to review
                </small>
              </div>
            </div>
            <div
              aria-label="Global"
              class="pf-c-menu"
              data-ouia-component-id="OUIA-Generated-Menu-2"
              data-ouia-component-type="PF4/Menu"
              data-ouia-safe="true"
            >
              <div
                class="pf-c-menu__content"
                style="--pf-c-menu__content--Height: 350px;"
              >
                <section
                  class="pf-c-menu__group"
                >
                  <h1
                    class="pf-c-menu__group-title"
                    id=""
                  >
                    Positive Tests
                  </h1>
                  <ul
                    class="pf-c-menu__list indentList"
                    id="positive-test-list"
                  >
                    <li
                      class="pf-c-menu__list-item pf-m-disabled"
                      tabindex="-1"
                    >
                      <button
                        aria-current="false"
                        class="pf-c-menu__item"
                        disabled=""
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="pf-c-menu__item-main"
                        >
                          <span
                            class="pf-c-menu__item-text"
                          >
                            [No Positive Tests]
                          </span>
                        </span>
                      </button>
                    </li>
                  </ul>
                </section>
                <section
                  class="pf-c-menu__group"
                >
                  <h1
                    class="pf-c-menu__group-title"
                    id=""
                  >
                    Certificates of Recovery
                  </h1>
                  <ul
                    class="pf-c-menu__list indentList"
                    id="recovery-list"
                  >
                    <li
                      class="pf-c-menu__list-item pf-m-disabled"
                      tabindex="-1"
                    >
                      <button
                        aria-current="false"
                        class="pf-c-menu__item"
                        disabled=""
                        tabindex="-1"
                        type="button"
                      >
                        <span
                          class="pf-c-menu__item-main"
                        >
                          <span
                            class="pf-c-menu__item-text"
                          >
                            [No Certificates of Recovery]
                          </span>
                        </span>
                      </button>
                    </li>
                  </ul>
                </section>
              </div>
            </div>
          </article>
        </div>
      </div>
    </div>
    <div
      class="pf-c-card__footer"
    >
      <div
        class="pf-l-flex"
      >
        <div
          class="pf-m-align-left"
        >
          <button
            aria-disabled="false"
            class="pf-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Button-plain-1"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            type="button"
          >
            Back
          </button>
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
